lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+  auto referenceDependenciesFile = readSupplementaryOutputArguments(
+      options::OPT_emit_reference_dependencies_path);
+  auto serializedDiagnostics =
+      readSupplementaryOutputArguments(options::OPT_serialize_diagnostics_path);
+  auto loadedModuleTrace = readSupplementaryOutputArguments(
+      options::OPT_emit_loaded_module_trace_path);
+  auto TBD = readSupplementaryOutputArguments(options::OPT_emit_tbd_path);
+
+  if (!objCHeaderOutput || !moduleOutput || !moduleDocOutput ||
+      !dependenciesFile || !referenceDependenciesFile ||
+      !serializedDiagnostics || !loadedModuleTrace || !TBD) {
+    return None;
+  }
+  std::vector<SupplementaryOutputPaths> result;
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  for (unsigned i = 0; i < N; ++i) {
+    result.push_back(SupplementaryOutputPaths(
+        (*objCHeaderOutput)[i], (*moduleOutput)[i], (*moduleDocOutput)[i],
+        (*dependenciesFile)[i], (*referenceDependenciesFile)[i],
+        (*serializedDiagnostics)[i], (*loadedModuleTrace)[i], (*TBD)[i]));
+  }
+  return result;
+}
+
+// Extend this routine for filelists if/when we have them.
+
+Optional<std::vector<std::string>>
+SupplementaryOutputPathsComputer::readSupplementaryOutputArguments(
+    swift::options::ID path_id) const {
+  std::vector<std::string> paths = Args.getAllArgValues(path_id);
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  if (paths.size() > N) {
+    std::unique_ptr<llvm::opt::OptTable> Table = createSwiftOptTable();
+    Diags.diagnose(SourceLoc(), diag::error_too_many_supplementary_outputs,
+                   Table->getOptionName(path_id));
+    return None;
+  }
+  while (paths.size() < N)
+    paths.push_back(std::string());
+  return paths;
+}
+
+Optional<SupplementaryOutputPaths>
+SupplementaryOutputPathsComputer::computeOutputPathsForOneInput(
+    StringRef outputFile, const SupplementaryOutputPaths &pathsFromArguments,
+    const InputFile &input) const {
+  StringRef implicitBasis = deriveImplicitBasis(outputFile, input);
+
+  using namespace options;
+
+  auto dependenciesFilePath = determineSupplementaryOutputFilename(
+      OPT_emit_dependencies, pathsFromArguments.DependenciesFilePath, "d", "",
+      implicitBasis);
+  if (!dependenciesFilePath)
+    return None;
+
+  auto referenceDependenciesFilePath = determineSupplementaryOutputFilename(
+      OPT_emit_reference_dependencies,
+      pathsFromArguments.ReferenceDependenciesFilePath, "swiftdeps", "",
+      implicitBasis);
+  if (!referenceDependenciesFilePath)
+    return None;
+
+  auto serializedDiagnosticsPath = determineSupplementaryOutputFilename(
+      OPT_serialize_diagnostics, pathsFromArguments.SerializedDiagnosticsPath,
+      "dia", "", implicitBasis);
+
+  if (!serializedDiagnosticsPath)
+    return None;
+
+  auto objCHeaderOutputPath = determineSupplementaryOutputFilename(
```

Comment:
Super nitpick: "ObjC" is one word when abbreviated as such, so its lowercase form is "objc".

Reply:
Fixed.